### PR TITLE
perf(node): reduce reflect FieldByName and Field calls

### PR DESCRIPTION
**Pull request by jayzhudev**
The `reflect` method calls are expensive to be put inside nested for-loops. The performance can be improved if the results of these calls are cached.

Here are some rough reference numbers from e2e testing sending 2000 leaf-list updates in a SetRequest:

| Version             | Time          |
| --------------- | ---------- |
| Before #829     | ~2,300 ms     | 
| **After** #829 | **~900 ms** |

### CODE REVIEW

File: ytypes/node.go [hunk 1]
```diff
@@ -341,6 +341,27 @@ func retrieveNodeList(schema *yang.Entry, root interface{}, path, traversedPath
 
 	listKeyT := rv.Type().Key()
 	listElemT := rv.Type().Elem()
+
+	listKeyFieldNameMap := map[int]string{}
+	schemaKeyMap := map[string]string{}
+
+	// Reduce FieldByName calls in the loop and reduce Field calls and tag lookup calls in the loop.
+	if listKeyT.Kind() == reflect.Struct {
```
Comment (wenovus) [inline ytypes/node.go:349]: util.IsTypeStruct(listKeyT)

File: ytypes/node.go [hunk 2]
```diff
@@ -341,6 +341,27 @@ func retrieveNodeList(schema *yang.Entry, root interface{}, path, traversedPath
...
+		for i := 0; i < listKeyT.NumField(); i++ {
+			fieldName := listKeyT.Field(i).Name
+			if elem, ok := listElemT.Elem().FieldByName(fieldName); ok {
+				schemaKey, err := directDescendantSchema(elem)
+				if err != nil {
+					return nil, status.Errorf(codes.Unknown, "unable to get direct descendant schema name for %v: %v", schemaKey, err)
+				}
+
+				schemaKeyMap[fieldName] = schemaKey
+			}
```
Comment (wenovus) [inline ytypes/node.go:359]: To keep the error statement close to where it was generated, you can return an error when ok=False like the original code
```go
if !ok {
				return nil, status.Errorf(codes.NotFound, "element struct type %v does not contain key field %s", listElemT, fieldName)
			}
```

File: ytypes/node.go [hunk 3]
```diff
@@ -407,22 +428,17 @@ func retrieveNodeList(schema *yang.Entry, root interface{}, path, traversedPath
...
 			fieldValue := k.Field(i)
 			if !fieldValue.IsValid() {
 				return nil, status.Errorf(codes.InvalidArgument, "invalid field %s in %T", fieldName, k)
 			}
 
-			elemFieldT, ok := listElemT.Elem().FieldByName(fieldName)
+			schemaKey, ok := schemaKeyMap[fieldName]
 			if !ok {
 				return nil, status.Errorf(codes.NotFound, "element struct type %v does not contain key field %s", listElemT, fieldName)
 			}
```
Comment (wenovus) [inline ytypes/node.go:440]: If you place this error statement above, then this error can be changed to say codes.Internal, and that it is expected that this field should always be populated.

**Review by wenovus — COMMENTED** [review]
A couple nits, otherwise looks good, thanks!

File: ytypes/node.go (hunk 3 above)
Comment (jayzhudev) [reply ytypes/node.go]: These are good suggestions, thanks. Changes made in the latest commit.

**Review by wenovus — APPROVED** [review]

**Pull request merged**